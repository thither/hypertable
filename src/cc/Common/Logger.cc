/*
 * Copyright (C) 2007-2016 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/** @file
 * Logging routines and macros.
 * The LogWriter provides facilities to write debug, log, error- and other
 * messages to stdout. The Logging namespaces provides std::ostream-
 * and printf-like macros and convenience functions.
 */

#include <Common/Compat.h>

#include "String.h"
#include "Logger.h"

#include <iostream>
#include <stdio.h>
#include <stdarg.h>
#include <mutex>

namespace Hypertable { namespace Logger {

static String logger_name;
static LogWriter *logger_obj = 0;
static std::mutex mutex;

void initialize(const String &name) {
  logger_name = name;
}

LogWriter *get() {
  if (!logger_obj)
    logger_obj = new LogWriter(logger_name);
  return logger_obj;
}

void LogWriter::log_string(int priority, const char *message) {
  static const char *priority_name[] = {
    "FATAL",
    "ALERT",
    "CRIT",
    "ERROR",
    "WARN",
    "NOTICE",
    "INFO",
    "DEBUG",
    "NOTSET"
  };

  std::lock_guard<std::mutex> lock(mutex);
  if (m_test_mode) {
    fprintf(m_file, "%s %s : %s\n", priority_name[priority], m_name.c_str(),
            message);
  }
  else {
    time_t t = ::time(0);
    fprintf(m_file, "%u %s %s : %s\n", (unsigned)t, priority_name[priority],
            m_name.c_str(), message);
  }

  flush();
}

void LogWriter::log_varargs(int priority, const char *format, va_list ap) {
  char buffer[1024 * 16];
  vsnprintf(buffer, sizeof(buffer), format, ap);
  log_string(priority, buffer);
}

void LogWriter::debug(const char *format, ...) {
  va_list ap;
  va_start(ap, format);
  log_varargs(Priority::DEBUG, format, ap);
  va_end(ap);
}

void LogWriter::log(int priority, const char *format, ...) {
  va_list ap;
  va_start(ap, format);
  log_varargs(priority, format, ap);
  va_end(ap);
}




int cfg::from_string(String loglevel) {
    if (loglevel == "info")
      return Logger::Priority::INFO;
    if (loglevel == "debug")
      return Logger::Priority::DEBUG;
    if (loglevel == "notice")
      return Logger::Priority::NOTICE;
    if (loglevel == "warn")
      return Logger::Priority::WARN;
    if (loglevel == "error")
      return Logger::Priority::ERROR;
    if (loglevel == "crit")
      return Logger::Priority::CRIT;
    if (loglevel == "alert")
      return Logger::Priority::ALERT;
    if (loglevel == "fatal")
      return Logger::Priority::FATAL;
    return -1;
}

String cfg::repr(int value) {
    switch(value){
      case Logger::Priority::INFO:
        return "info";
      case Logger::Priority::DEBUG:
        return "debug";
      case Logger::Priority::NOTICE:
        return "notice";
      case Logger::Priority::WARN:
        return "warn";
      case Logger::Priority::ERROR:
        return "error";
      case Logger::Priority::CRIT:
        return "crit";
      case Logger::Priority::ALERT:
        return "alert";
      case Logger::Priority::FATAL:
        return "fatal";
      default:
        return format("undefined logging level: %d", value);
    }
}

}} // namespace Hypertable::
