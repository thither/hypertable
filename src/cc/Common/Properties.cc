/*
 * Copyright (C) 2007-2016 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/** @file
 * Program options handling.
 * Based on boost::program_options, the options are configured in Config.h.
 */

#include <Common/Compat.h>
#include <Common/Properties.h>

#include <errno.h>
#include <fstream>

using namespace Hypertable;


void
Properties::load(const String &fname, PropertiesDesc &desc,
                 bool allow_unregistered) {
  std::ifstream in(fname.c_str());
  Config::Parser prs(in, desc, allow_unregistered);
  load_from(prs.get_options());
} 

void
Properties::load_files_by(const String &names_prop, PropertiesDesc &desc, 
                            bool allow_unregistered) {
  if(names_prop.empty() || !has(names_prop)) 
    return;

  Strings files = get_strs(names_prop);
  for (Strings::iterator it=files.begin(); it<files.end(); it++){
	  try {
      load(*it, desc, allow_unregistered);
    }
	  catch (std::exception &e) {
		  HT_WARNF("%s has bad cfg file %s: %s", 
                names_prop.c_str(), (*it).c_str(), e.what());
    }
  }
}

String 
Properties::reload(const String &fname, PropertiesDesc &desc, 
                   bool allow_unregistered) {
  String out;
	try {
		std::ifstream in(fname.c_str());

    HT_INFOF("Reloading Configuration File %s", fname.c_str());
		if (!in) throw;
    
    out.append("\n\nCurrent Configurations:\n");
    append_to_string(out);

    Config::Parser prs(in, desc, allow_unregistered);
    load_from(prs.get_options(), true);

    out.append("\n\nNew Configurations:\n");
    append_to_string(out);
    return out;
	}
	catch (std::exception &e) {
		HT_WARNF("Error::CONFIG_BAD_CFG_FILE %s: %s", fname.c_str(), e.what());
    return format("Error::CONFIG_BAD_CFG_FILE %s: %s \noutput:\n%s", 
                    fname.c_str(), e.what(), out.c_str());
	}
}

void
Properties::parse_args(int argc, char *argv[], PropertiesDesc &desc,
                       PropertiesDesc *hidden, bool allow_unregistered) {
  
  Config::Parser prs(argc, argv, desc, hidden, allow_unregistered);
  load_from(prs.get_options());
}

void
Properties::parse_args(const std::vector<String> &args,
                      PropertiesDesc &desc, PropertiesDesc *hidden,
                       bool allow_unregistered) {
  Config::Parser prs(args, desc, hidden, allow_unregistered);
  load_from(prs.get_options());
}
 
void 
Properties::load_from(Config::Parser::Options opts, bool only_guarded) {
  AliasMap::iterator it;
  for(const auto &kv : opts){
    if(has(kv.first) && (kv.second->is_default() 
                          || (only_guarded && !kv.second->is_guarded())))
      continue;

    set(kv.first, kv.second);
      
    it = m_alias_map.find(kv.first);
    if(it != m_alias_map.end())
      set((*it).second, get_value_ptr(kv.first));
  }
}

void Properties::alias(const String &primary, const String &secondary) {
  m_alias_map[primary] = secondary;
  m_alias_map[secondary] = primary;
}

void
Properties::print(std::ostream &out, bool include_default) {
  for (const auto &kv : m_map) {
    bool isdefault = kv.second->is_default();

    if (include_default || !isdefault) {
      out << kv.first << '=' << kv.second->str();

      if (isdefault)
        out << " (default)";
      out << std::endl;
    }
  }
}

void
Properties::append_to_string(String &out, bool include_default) {
  for (const auto &kv : m_map) {
    bool isdefault = kv.second->is_default();

    if (include_default || !isdefault) {
      out.append(format("%s=%s", kv.first.c_str(), kv.second->str().c_str()));

      if (isdefault)
        out.append(" (default)");
      out.append("\n");
    }
  }
}

