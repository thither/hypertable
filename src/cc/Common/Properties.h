/*
 * Copyright (C) 2007-2016 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

/** @file
 * Program options handling.
 * Based on boost::program_options, the options are configured in Config.h.
 */

#ifndef Common_Properties_h
#define Common_Properties_h

#include <vector>
#include <string>
#include <mutex>
#include <Common/Error.h>
#include <Common/Logger.h>
#include <Common/Property.h>
#include <Common/PropertiesParser.h>


// convenience/abbreviated accessors
#define HT_PROPERTIES_ABBR_ACCESSORS() \
  inline bool get_bool(const String &name)  { \
    return get<bool>(name); } \
  inline String get_str(const String &name)  { \
    return get<String>(name); } \
  inline Strings get_strs(const String &name)  { \
    return get<Strings>(name); } \
  inline uint16_t get_i16(const String &name)  { \
    return get<uint16_t>(name); } \
  inline int32_t get_i32(const String &name)  { \
    return get<int32_t>(name); } \
  inline int64_t get_i64(const String &name)  { \
    return get<int64_t>(name); } \
  inline Int64s get_i64s(const String &name)  { \
    return get<Int64s>(name); } \
  inline double get_f64(const String &name)  { \
    return get<double>(name); } \
  inline Doubles get_f64s(const String &name)  { \
    return get<Doubles>(name); } \
  inline bool get_bool(const String &name, bool default_value) \
     { return get<bool>(name, default_value); } \
  inline String get_str(const String &name, String &default_value) \
     { return get<String>(name, static_cast<String>(default_value)); } \
  inline String get_str(const String &name, String default_value) \
     { return get<String>(name, default_value); } \
  inline Strings get_strs(const String &name, Strings default_value) \
     { return get<Strings>(name, default_value); } \
  inline uint16_t get_i16(const String &name, uint16_t default_value) \
     { return get<uint16_t>(name, default_value); } \
  inline int32_t get_i32(const String &name, int32_t default_value) \
     { return get<int32_t>(name, default_value); } \
  inline int64_t get_i64(const String &name, int64_t default_value) \
     { return get<int64_t>(name, default_value); } \
  inline Int64s get_i64s(const String &name, Int64s &default_value) \
     { return get<Int64s>(name, default_value); } \
  inline double get_f64(const String &name, double default_value) \
     { return get<double>(name, default_value); } \
  inline Doubles get_f64s(const String &name, Doubles default_value) \
     { return get<Doubles>(name, default_value); }
  /*     \
  inline bool get_bool(const String &name, bool default_value) \
    const { return get<bool>(name, default_value); } \
  inline String get_str(const String &name, String default_value) \
    const { return get<String>(name, default_value); } \
  inline Strings get_strs(const String &name, Strings &default_value) \
    const { return get<Strings>(name, default_value); } \
  inline uint16_t get_i16(const String &name, uint16_t default_value) \
    const { return get<uint16_t>(name, default_value); } \
  inline int32_t get_i32(const String &name, int32_t default_value) \
    const { return get<int32_t>(name, default_value); } \
  inline int64_t get_i64(const String &name, int64_t default_value)  \
    const { return get<int64_t>(name, default_value); } \
  inline Int64s get_i64s(const String &name, Int64s default_value) \
    const { return get<Int64s>(name, default_value); } \
  inline double get_f64(const String &name, double default_value) \
    const { return get<double>(name, default_value); } \
  inline Doubles get_f64s(const String &name, Doubles default_value) \
    const { return get<Doubles>(name, default_value); }
  */
 

namespace Hypertable {


/** @addtogroup Common
 *  @{
 */

typedef Config::ParserConfig PropertiesDesc;


/**
 * Manages a collection of program options.
 *
 * Implements getters and setters, aliases and default values.
 */
class Properties {
  typedef std::map<String, Property::ValuePtr> Map;
  typedef std::pair<String, Property::ValuePtr> MapPair;
  typedef std::pair<Map::iterator, bool> InsRet;
  typedef std::map<String, String> AliasMap;
  
public:
  /** Default constructor; creates an empty set */
  Properties() {  }

  /** Constructor; load properties from a filename
   *
   * @param filename The name of the file with the properties
   * @param desc A Property description with valid properties, default values
   * @param allow_unregistered If true, unknown/unregistered properties are
   *        accepted
   */
  Properties(const String &filename, PropertiesDesc &desc,
             bool allow_unregistered = false) {
    load(filename, desc, allow_unregistered);
  }

  /**
   * Loads the parsed options to properties
   *
   * @param opts onfig::Parser::Options
   * @param only_guarded update only huarded property
   */
  void load_from(Config::Parser::Options opts, bool only_guarded = false);

  /**
   * Loads a configuration file with properties
   *
   * @param filename The name of the configuration file
   * @param desc A property description
   * @param allow_unregistered If true, unknown/unregistered properties are
   *        accepted
   */
  void load(const String &filename, PropertiesDesc &desc,
	  bool allow_unregistered = false);

  /**
   * Loads a configuration files by property name
  *
   * @param names_prop The name of property name
   * @param desc A property description
   * @param allow_unregistered If true, unknown/unregistered properties are
   *        accepted
   */
  void load_files_by(const String &names_prop, PropertiesDesc &desc, 
	  bool allow_unregistered = false);
    
  /**
   * ReLoads a configuration file with properties
   *
   * @param filename The name of the configuration file
   * @param desc A property description
   * @param allow_unregistered If true, unknown/unregistered properties are
   *        accepted
   */
  String reload(const String &filename, PropertiesDesc &desc,
	  bool allow_unregistered = false);

  /**
   * Parses command line arguments
   *
   * @param argc The argument count (from main)
   * @param argv The argument array (from main)
   * @param desc The options description
   * @param hidden The hidden options description
   * @param allow_unregistered If true, unknown/unregistered properties are
   *        accepted
   * @throw Error::CONFIG_INVALID_ARGUMENT on error
   */
  void parse_args(int argc, char *argv[], PropertiesDesc &desc,
              PropertiesDesc *hidden = 0, bool allow_unregistered = false);

  /**
   * Parses command line arguments
   *
   * @param args A vector of Strings with the command line arguments
   * @param desc The options description
   * @param hidden The hidden options description
   * @param allow_unregistered If true, unknown/unregistered properties are
   *        accepted
   * @throw Error::CONFIG_INVALID_ARGUMENT on error
   */
  void parse_args(const std::vector<String> &args, PropertiesDesc &desc,
              PropertiesDesc *hidden = 0, bool allow_unregistered = false);

  /**
   * Get the ptr of property value. Throws if name is not defined.
   * 
   * @param name The name of the property
   * @throw Error::CONFIG_GET_ERROR if the requested property is not defined
   * @return Property::ValuePtr
   */
  Property::ValuePtr get_value_ptr(const String &name) {
    try{
      return m_map.at(name);
    }
    catch (std::exception &e) {
      HT_THROWF(Error::CONFIG_GET_ERROR, "getting value of '%s': %s",
                name.c_str(), e.what());
    }
    // 
    // return nullptr;
  }

  /**
   * Get the ptr of property value type. Throws if name is not defined.
   * Property::ValueDef<int32_t>* threads = props->get_type_ptr<int32_t>("threads");
   * 
   * @param name The name of the property
   * @throw Error::CONFIG_GET_ERROR if the requested property is not defined
   * @return Property::ValueDef<T>*
   */
  template <typename T>
  Property::ValueDef<T>* get_type_ptr(const String &name) {
    return (Property::ValueDef<T>*)get_value_ptr(name)->get_type_ptr();
  }

  /**
   * Get the property value ptr
   *
   * @param name The name of the property
   * @throw Error::CONFIG_GET_ERROR if the requested property is not defined
   * @return T*
   */
  template <typename T>
  T* get_ptr(const String &name) {
    return get_value_ptr(name)->get_ptr<T>();
  }
  
  /**
   * Get the String representation of property name 
   *
   * @param name The name of the property
   * @throw Error::CONFIG_GET_ERROR if the requested property is not defined
   * @return const String
   */
  const String str(const String &name) {
    try {
      return get_value_ptr(name)->str();
    }
    catch (std::exception &e) {
      HT_THROWF(Error::CONFIG_GET_ERROR, "getting value of '%s': %s",
                name.c_str(), e.what());
    }
  }
  /**
   * Get the value of option of type T. Throws if option is not defined.
   *
   * @param name The name of the property
   * @throw Error::CONFIG_GET_ERROR if the requested property is not defined
   */
  template <typename T>
  T get(const String &name) {
    try {
      return get_value_ptr(name)->get<T>();
    }
    catch (std::exception &e) {
      HT_THROWF(Error::CONFIG_GET_ERROR, "getting value of '%s': %s",
                name.c_str(), e.what());
    }
  }
  
  /**
   * Get the value of option of type T by preference. Throws if option is not defined.
   *
   * @param namee The names of preference for the property
   * @throw Error::CONFIG_GET_ERROR if the requested property is not defined
   */
  template <typename T>
  T get_pref(Strings names){
    for(String name : names)
      if (has(name))
        return get<T>(name);
    HT_THROWF(Error::CONFIG_GET_ERROR, "getting pref value by '%s'",
                                        format_list(names).c_str());
  }

  /**
   * Get the value of option of type T. Returns supplied default value if
   * not found. Try use the first form in usual cases and supply default
   * values in the config descriptions, as it validates the name and is less
   * error prone.
   * If no value, sets the default value to properties 
   *
   * @param name The name of the property
   * @param default_value The default value to return if not found
   */
  template <typename T>
  T get(const String &name, T default_value) {
    try {
      Map::const_iterator it = m_map.find(name);

      if (it != m_map.end())
        return (*it).second->get<T>();
      
      set(name, default_value);
    
      return get<T>(name);
    }
    catch (std::exception &e) {
      HT_THROWF(Error::CONFIG_GET_ERROR, "getting value of '%s': %s",
                name.c_str(), e.what());
    }
  }
  /*
  template <typename T>
  const T get(const String &name, T &default_value)  {
    T v = get(name, default_value);
    return std::as_const(v);
  }
  */

  /**
   * Check whether a property is by default value
   *
   * @param name The name of the property
   * @return true if the value is default
   */
  bool defaulted(const String &name) {
    return get_value_ptr(name)->is_default();
  }

  /** Check whether a property exists
   *
   * @param name The name of the property
   * @return true if the property exists
   */
  bool has(const String &name) {
    return m_map.count(name);
  }
  /*
  bool has(const String &name) const {
    return m_map.count(name);
  }
  */

  HT_PROPERTIES_ABBR_ACCESSORS()

  /**
   * Add property to the map of type T
   *
   * @param name The name of the property
   * @param v The Property::ValueDef<T> the property
   * @return An iterator to the new item
   */
  template<typename T>
  InsRet add(const String &name, T v) {
    return m_map.insert(MapPair(name, new Property::Value(v)));
  }
  /**
   * Add property to the map
   *
   * @param name The name of the property
   * @param v The Property::ValuePtr of the property
   * @return An iterator to the new item
   */
  InsRet add(const String &name, Property::ValuePtr v) {
    return m_map.insert(MapPair(name, v));
  }

  /**
   * Set a property in the map, create or update, of T type
   *
   * @param name The name of the property
   * @param v The value of the property, name need to correspond to Value Type
   */
  template<typename T>
  void set(const String &name, T v, bool defaulted=false)  {
    InsRet r = add(name, v);
    if (!r.second){
      (*r.first).second = new Property::Value(v);
    } else 
      (*r.first).second->set_value(v);
    if(defaulted)
      (*r.first).second->default_value();
  }

  /**
   * Set a property in the map, create or update, by Property::ValueDef<T> type
   *
   * @param name The name of the property
   * @param v The Property::ValueDef<T>
   */
  template<typename T>
  void set(const String &name, Property::ValueDef<T> v)  {
    InsRet r = add(name, v.get_value());
    if (!r.second){
      (*r.first).second = new Property::Value(v.get_value());
    } else 
      (*r.first).second->set_value(v.get_value());
  }

  /**
   * Set a property in the map, create or update, from Property::ValuePtr
   *
   * @param name The name of the property
   * @param v The Property::ValuePtr
   */
  void set(const String &name, Property::ValuePtr p) {
    Property::ValuePtr p_set;
    if(m_map.count(name)==0){
      p_set = Property::make_new(p);
      
      if(p->is_default())
        p_set->default_value();
      
      if(p->is_guarded())
        p_set->guarded(true);
        
      add(name, p_set);

    } else if(!p->is_default()) {
      p_set = get_value_ptr(name);
      p_set->set_value_from(p);
      p_set->default_value(false);
    }
  }

  /**
   * Remove a property from the map
   *
   * @param name The name of the property
   */
  void remove(const String &name) {
    m_map.erase(name);
  }

  /**
   * Setup an alias for a property.
   *
   * The primary property has higher priority, meaning when
   * aliases are sync'ed the primary value can override secondary value
   *
   * @param primary The primary property name
   * @param secondary The secondary property name
   */
  void alias(const String &primary, const String &secondary);

  /**
   * Returns all property names
   *
   * @param names reference to vector to hold names of all properties
   */
  void get_names(std::vector<String> &names) {
    for (Map::const_iterator it = m_map.begin(); it != m_map.end(); it++)
      names.push_back((*it).first);
  }

  /**
   * Prints keys and values of the configuration map
   *
   * @param out The output stream
   * @param include_default If true then default values are included
   */
  void print(std::ostream &out, bool include_default = false);

  /**
   * Append to String keys and values of the configuration map for print
   *   
   * @param out The String to append 
   * @param include_default If true then default values are included
   */
  void append_to_string(String &out, bool include_default = false);

private:

  /** The map containing all properties */
  Map m_map;
  
  /** A map with all aliases */
  AliasMap m_alias_map;
};

 typedef std::shared_ptr<Properties> PropertiesPtr;



/** Helper class to access parts of the properties.
 *
 * This snippet extracts all properties that start with
 * "Hypertable.RangeServer." and uses the SubProperty to access
 * "Hypertable.RangeServer.Range.SplitSize". 
 *
 *     SubProperties cfg(props, "Hypertable.RangeServer.");
 *     i64_t foo = cfg.get_i64("Range.SplitSize");
 */
class SubProperties {
public:
  /** Constructor
   *
   * @param props The original Properties object
   * @param prefix The common prefix of all sub-properties
   */
  SubProperties(PropertiesPtr &props, const String &prefix)
    : m_props(props), m_prefix(prefix) {
  }

  /** Returns the full name of a sub-property
   *
   * @param name The name of the sub-property
   * @return The full name (prefix + name)
   */
  String full_name(const String &name) const {
    return m_prefix + name;
  }

  /**
   * Calls Properties::get for a sub-property.
   *
   * @param name The name of the sub-property
   */
  template <typename T>
  T get(const String &name) {
    return m_props->get<T>(full_name(name));
  }

  /**
   * Calls Properties::get for a sub-property.
   *
   * @param name The name of the sub-property
   * @param default_value The default value to return if not found
   */
  template <typename T>
  T get(const String &name, T default_value) {
    return m_props->get<T>(full_name(name), default_value);
  }

  /**
   * Check whether a sub-property has a default value
   *
   * @param name The name of the sub-property
   * @return true if the value is default
   */
  bool defaulted(const String &name) {
    return m_props->defaulted(full_name(name));
  }

  /**
   * Check whether a sub-property exists
   *
   * @param name The name of the sub-property
   * @return true if the property exists
   */
  bool has(const String &name) {
    return m_props->has(full_name(name));
  }

  HT_PROPERTIES_ABBR_ACCESSORS()

private:
  /** Reference to the original Properties object */
  PropertiesPtr m_props;

  /** The common prefix of all sub-properties */
  String m_prefix;
};

/** @} */

}
#endif // Common_Properties_h
