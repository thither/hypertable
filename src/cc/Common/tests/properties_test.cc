/** -*- C++ -*-
 * Copyright (C) 2007-2016 Hypertable, Inc.
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Hypertable. If not, see <http://www.gnu.org/licenses/>
 */

#include "Common/Compat.h"
#include "Common/Logger.h"
#include "Common/Properties.h"
#include <iostream>

using namespace Hypertable;
using namespace Property;
using namespace std;

namespace {

enum Mode { MODE_FOO, MODE_BAR };

void basic_test(PropertiesDesc &desc) {
  static const char *argv[] = { "test", "--string", "foo", "--strs", "s1",
      "--i16", "64k", "--i32", "20M", "--i64", "8g", "--strs", "s2",
      "--int64s", "16G", "--int64s", "32G" };
  vector<String> args;
  args.push_back("--f64");      args.push_back("1k");
  args.push_back("--float64s"); args.push_back("1.5");
  args.push_back("--float64s"); args.push_back("1.5K");
  args.push_back("--float64s"); args.push_back("1.5M");
  args.push_back("--float64s"); args.push_back("1.5g");

  Properties props;
  props.parse_args(sizeof(argv) / sizeof(char *), (char **)argv, desc);
  props.parse_args(args, desc);
  //props.set("string_type", (String)"11111 a String Type 111111");
  //cout << "string_type=" << props.get<String>("string_type") << endl;
  
  props.set("mode", MODE_FOO);
  cout << "mode=" << props.str("mode") << endl;
  props.print(cout, true);
}
/*
void notify(const String &s) {
  cout << __func__ <<": "<< s << endl;
}
*/
}

int main(int argc, char *argv[]) {
  try {
    double f64test = 0;
    PropertiesDesc desc;
    desc.add_options()
      ("help,h", "help")
      ("boo", boo(true)->zero_token(), "a boolean arg")
      ("string", str(), "a string arg")  // ->notifier(notify) >> onChange cb
      ("strs", strs(), "a list of strings")
      ("i16", i16(), "a 16-bit integer arg")
      ("i32", i32(), "a 32-bit integer arg")
      ("i64", i64(), "a 64-bit integer arg")
      ("int64s", i64s(), "a list of 64-bit integers")
      ("f64", f64(f64test), "a double arg")
      ("float64s", f64s(), "a list of doubles")
      ;
    Properties props;
    props.parse_args(argc, argv, desc);

    if (props.has("help"))
      cout << desc << endl;

    cout << "f64test=" << f64test << endl;
    /*
    if (!props.has("f64")) {
      cout << "f64 has!=" << f64test << endl;
      props.set("f64", (double)1234.1234);
      cout << "f64=" << props.str("f64") << endl;
    } else
      cout << "f64 has=" << f64test << endl;
    props.get_value_ptr("f64")->set_value((double)5678.9);
    cout << "f64 set=" << props.str("f64") << endl;
    */

    props.print(cout);
    HT_TRY("basic test", basic_test(desc));
  }
  catch (Exception &e) {
    cout << e << endl;
    return 1;
  }
  return 0;
}
