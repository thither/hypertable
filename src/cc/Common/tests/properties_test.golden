f64test=0
mode=An ENUM TYPE
boo=true (default)
f64=1000
float64s=[1.5, 1500, 1.5e+06, 1.5e+09]
i16=64000
i32=20000000
i64=8000000000
int64s=[16000000000, 32000000000]
mode=An ENUM TYPE
string=foo
strs=[s1, s2]
